File: 0034-find-first-and-last-position-of-element-in-sorted-array/Solution.cs
public class Solution {
    public int[] SearchRange(int[] nums, int target)
    {
        var firstIndex = FindFirst(nums, target);
        var lastIndex = FindLast(nums, target);
        if (firstIndex <= lastIndex && lastIndex < nums.Length && nums[firstIndex] == target && nums[lastIndex] == target)
        {
            return new int[] { firstIndex, lastIndex };
        }
        else
        {
            return new int[] { -1, -1 };
        }
    }

    private int FindFirst(int[] nums, int target)
    {
        var leftIndex = 0;
        var rightIndex = nums.Length - 1;

        while (leftIndex <= rightIndex)
        {
            var middleIndex = (leftIndex + rightIndex) / 2;
            if (nums[middleIndex] < target)
            {
                leftIndex = middleIndex + 1;
            }
            else
            {
                rightIndex = middleIndex - 1;
            }
        }
        return leftIndex;
    }

    private int FindLast(int[] nums, int target)
    {
        var leftIndex = 0;
        var rightIndex = nums.Length - 1;

        while (leftIndex <= rightIndex)
        {
            var middleIndex = (leftIndex + rightIndex) / 2;
            if (nums[middleIndex] <= target)
            {
                leftIndex = middleIndex + 1;
            }
            else
            {
                rightIndex = middleIndex - 1;
            }
        }
        return rightIndex;
    }

}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/Solution.go
package Leetcode

func searchRange(nums []int, target int) []int {
	first := findFirst(nums, target)
	last := findLast(nums, target)
	if first <= last && last < len(nums) && nums[first] == target && nums[last] == target {
		return []int{first, last}
	} else {
		return []int{-1, -1}
	}
}

func findFirst(nums []int, target int) int {
	left := 0
	right := len(nums) - 1

	for left <= right {
		mid := (left + right) / 2
		if nums[mid] >= target {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return left
}

func findLast(nums []int, target int) int {
	left := 0
	right := len(nums) - 1
	for left <= right {
		mid := (left + right) / 2
		if nums[mid] <= target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return right
}
